src/core/inductive.ml

```diff
@@ -0,0 +1,244 @@
+  let p = Bindlib.new_var mkfree p_str in
+  let app  : tbox -> tbox list -> tbox = List.fold_left _Appl in
+  let fapp : sym  -> tbox list -> tbox = fun f ts -> app (_Symb f) ts in
+  let prf_of_p : tbox list -> tbox -> tbox = fun ts t ->
+    fapp c.symb_prf [_Appl (app (_Vari p) ts) t]
+  in
+
+  (* STEP 1: Create the type of the property p *)
+  let prop = _Symb c.symb_Prop in
+  let codom ts = _Impl (fapp ind_sym ts) prop in
+  let p_type = gen_ind_typ_codom pos ind_sym codom x_str in
+
+  (* STEP 2: Create each clause according to a constructor *)
+  (* [case_of cons_sym] creates a clause according to a constructor
+     [cons_sym]. *)
+  let case_of : sym -> tbox = fun cons_sym ->
+    let codom : term list -> tvar list -> tbox -> tbox = fun ts xs _ ->
+      prf_of_p (List.map lift ts) (fapp cons_sym (List.rev_map _Vari xs))
+    in
+    let inj_var : tvar list -> tvar -> tvar = fun _ x -> x in
+    let build_rec_hyp : term list -> tvar -> tbox = fun ts x ->
+      prf_of_p (List.map lift ts) (_Vari x)
+    in
+    let domrec : term -> tvar -> tbox -> tbox -> tbox =
+      fun a x rec_hyp next ->
+      _Prod (lift a) (Bindlib.bind_var x (_Impl rec_hyp next))
+    in
+    let dom : term -> tvar -> tbox -> tbox = fun a x next ->
+      _Prod (lift a) (Bindlib.bind_var x next)
+    in
+    let f_rec : tbox -> tvar -> tbox -> tbox = fun a _ _ -> a in
+    let f_not_rec : tbox -> tvar -> tbox = fun a _ -> a in
+    let init : tbox = _Type in
+    fold_cons_typ pos codom inj_var build_rec_hyp domrec dom ind_sym
+      cons_sym f_rec f_not_rec init x_str
+  in
+
+  (* STEP 3: Create the conclusion of the inductive principle *)
+  let codom ts =
+    let x = Bindlib.new_var mkfree x_str in
+    let t = Bindlib.bind_var x (prf_of_p ts (_Vari x)) in
+    _Prod (fapp ind_sym ts) t
+  in
+  let conclusion = gen_ind_typ_codom pos ind_sym codom x_str in
+
+  (* STEP 4: Create the induction principle *)
+  let t =
+    List.fold_right (fun a b -> _Impl (case_of a) b) cons_list conclusion
+  in
+  let t = _Prod p_type (Bindlib.bind_var p t) in
+  if !log_enabled then
+    log_ind "The induction principle of the inductive type [%a] is %a"
+      Pretty.pp_ident (Pos.none ("ind_"^ind_sym.sym_name))
+      Print.pp_term (Bindlib.unbox t);
+  Bindlib.unbox t
+
+(** [gen_rec_rules pos ind_sym rec_sym cons_list] returns the p_rules associa-
+    ted with an induction principle [rec_sym] of the inductive type [ind_sym]
+    (with its constructors [cons_list]). That means the command
+    inductive T : Π(i1:A1),...,Π(im:Am), TYPE := c1:T1 | ... | cn:Tn generates
+    a rule for each constructor. If Ti = Π(b1:B1), ... , Π(bk:Bk), T then the
+    rule for ci is:
+    rule ind_T p pic1 ... picn _ ... _ (ci x0 ... x(k-1)) -->
+    pici x0 t0? ... x(k-1) t(k-1)?
+    with m underscores and tj? = nothing if xj isn't a value of the inductive
+    type and tj? = (ind_T p pic1 ... picn _ ... _ xj) if xj is a value of the
+    inductive type T (i.e. xj = T v1 ... vm) *)
```

Comment:
Thanks to add a remark saying that there cannot be name clashes between pattern variable names and function symbols names since pattern variables are prefixed by `$`.